KEYCLOAK-2474: As Keycloak user I would like to be able to extend the data model

ewjmulder opened the pull request:
A. Add custom JPA classes to the Keycloak EntityManager through an SPI
B. (more easily) Extend the Liquibase database definition with custom
extra tables
C. Add extra resources to the Keycloak REST API through an SPI
D. Be able to lookup your own custom SPI's (when deployed in modules)
through the Keycloak session

ewjmulder committed:
KEYCLOAK-2474: As Keycloak user I would like to be able to extend the
data model
A. Add custom JPA classes to the Keycloak EntityManager through an SPI
B. (more easily) Extend the Liquibase database definition with custom
extra tables
C. Add extra resources to the Keycloak REST API through an SPI
D. Be able to lookup your own custom SPI's (when deployed in modules)
through the Keycloak session

ewjmulder commented:
As discussed over the mailing list on various occasions and in the JIRA ticket. The changes are meant to be low impact and not changing any existing Keycloak functionality. (Eclipse 'organize imports' changed the import sections a little, let me know if this is undesired)

I know this is too late for 1.9.x and is meant for 2.x. But I'd appreciate a review / some feedback earlier, so I know whether this is ok or I need to make some changes.

If the PR is ok, I will write a section about this new functionality to add to the documentation pages.

stianst commented:
Sorry for not giving you any feedback on this yet, but we've been quite busy the last week. Once 1.9.0.Final is out I'll take a look at it and give you some feedback. Thanks for sending this PR.

stianst commented:
DO NOT MERGE UNTIL 1.9.x IS CREATED

ewjmulder commented:
Any update / feedback on this PR?

stianst commented:
Sorry for the delay. I will review and merge as soon as I get the chance.
Will probably be a few weeks though. As we're still working on 1.9.x we
need to prioritise that at the moment.
On 17 Mar 2016 11:45, "Erik Mulder" <notifications@github.com> wrote:

> Any update / feedback on this PR?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/keycloak/keycloak/pull/2219#issuecomment-197814929>
>

stianst commented on services/src/main/java/org/keycloak/services/resources/RealmsResource.java:
```diff
@@ -183,4 +189,31 @@ public Response getWellKnown(final @PathParam("realm") String name,
         return Response.ok(wellKnown.getConfig()).cacheControl(CacheControlUtil.getDefaultCacheControl()).build();
     }
 
+    /**
```
Please remove adding custom REST endpoints. It's beyond the scope of this work and we have someone else working on it already.

stianst commented on services/src/main/java/org/keycloak/services/resources/admin/RealmAdminResource.java:
```diff
@@ -773,4 +775,28 @@ public void clearUserCache() {
+     * @return the rest resource or null if no implementation found
+     */
+    @Path("{unmapped_path}")
+    public Object resolveUnknowPath(@PathParam("unmapped_path") String path) {
```
Please remove adding custom REST endpoints. It's beyond the scope of this work and we have someone else working on it already.

stianst commented:
I've had a quick look through this. It looks OK to me, but I'm going to assign someone from the team to review it properly.

One thing the custom REST endpoints feature needs to be removed. It's beyond the scope of this issue and we already have someone working on that feature.

ewjmulder commented:
Thanks for the quick review. I included the custom REST endpoints feature, because the data model extension is not very useful without a way to manipulate the data using some kind of external interface, like a REST API. But if this will be included in another way, that works as well of course. I'll remove that part and try to resync it with the current state of the master.

ewjmulder commented:
@mposolda: I am updating the PR now and will make these changes / merges:
- Remove REST extensions
- Resolve conflicts around Liquibase changeset customization, using the new LiquibaseConnectionProvider
I will post a new / updated PR today.

ewjmulder committed:
Merge remote-tracking branch 'upstream/master'

Conflicts:
	model/jpa/src/main/java/org/keycloak/connections/jpa/DefaultJpaConnectionProviderFactory.java
	model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/LiquibaseJpaUpdaterProvider.java
	model/jpa/src/main/resources/META-INF/services/org.keycloak.provider.Spi
	services/src/main/java/org/keycloak/services/resources/RealmsResource.java
	services/src/main/java/org/keycloak/services/resources/admin/RealmAdminResource.java

ewjmulder committed:
Added newlines at the end of these files, for consistency.

ewjmulder commented:
Ok, PR updated. Ready for further review.

mposolda commented:
The PR looks good to me, Thanks! Few minor things:

- You removed custom REST endpoints, but you still kept the possibility to load SPI definitions from custom locations (the point D from your JIRA - the changes inside classes DefaultKeycloakSessionFactory, ProviderLoader, ProviderManager, DefaultProviderLoader). To me, it looks good how you implemented it, on the other hand it doesn't look very useful without the pluggable REST endpoints. I am fine with keep that in the PR, but not sure if it not clashes with the work someone else is doing on pluggable REST endpoints @stianst what do you think?

- Missing documentation
- Missing testing (Actually we prefer to add new tests to testsuite/integration-arquillian subproject)
- The example would be cool as well (Likely inside examples/providers ). IMO it makes sense to document and add an example especially for the custom JPA classes together with extending liquibase database definition (points A and B from your JIRA). Point D will likely deserve separate example later once we have C, so we will create another separate example for custom REST endpoints and custom SPI definition.

One question: Is extending liquibase work with updates? My point is, if you create custom liquibase file like jpa-changelog-master2.xml and then you need to update keycloak version from 1.9 to 2.0 , do you need to manually upgrade your custom jpa-changelog-master2.xml file too and include the new changelog file jpa-changelog-2.0.0.xml, which was added to keycloak in the meantime? If yes, we likely shouldn't use it in example (IMO we shouldn't have another manual step for maintain the example and manually update it during each keycloak release).

stianst commented:
I'm happy with the changes inside classes DefaultKeycloakSessionFactory,
ProviderLoader, ProviderManager, DefaultProviderLoader).

It would be better if it added additional liquibase changelog, rather than
rely on modifying jpa-changelog-master.xml. Should be fairly easy to do.



On 22 March 2016 at 08:08, Marek Posolda <notifications@github.com> wrote:

> The PR looks good to me, Thanks! Few minor things:
>
>    -
>
>    You removed custom REST endpoints, but you still kept the possibility
>    to load SPI definitions from custom locations (the point D from your JIRA -
>    the changes inside classes DefaultKeycloakSessionFactory, ProviderLoader,
>    ProviderManager, DefaultProviderLoader). To me, it looks good how you
>    implemented it, on the other hand it doesn't look very useful without the
>    pluggable REST endpoints. I am fine with keep that in the PR, but not sure
>    if it not clashes with the work someone else is doing on pluggable REST
>    endpoints @stianst <https://github.com/stianst> what do you think?
>    -
>
>    Missing documentation
>    - Missing testing (Actually we prefer to add new tests to
>    testsuite/integration-arquillian subproject)
>    - The example would be cool as well (Likely inside examples/providers
>    ). IMO it makes sense to document and add an example especially for the
>    custom JPA classes together with extending liquibase database definition
>    (points A and B from your JIRA). Point D will likely deserve separate
>    example later once we have C, so we will create another separate example
>    for custom REST endpoints and custom SPI definition.
>
> One question: Is extending liquibase work with updates? My point is, if
> you create custom liquibase file like jpa-changelog-master2.xml and then
> you need to update keycloak version from 1.9 to 2.0 , do you need to
> manually upgrade your custom jpa-changelog-master2.xml file too and include
> the new changelog file jpa-changelog-2.0.0.xml, which was added to keycloak
> in the meantime? If yes, we likely shouldn't use it in example (IMO we
> shouldn't have another manual step for maintain the example and manually
> update it during each keycloak release).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/keycloak/keycloak/pull/2219#issuecomment-199688847>
>

mposolda commented:
>> I'm happy with the changes inside classes DefaultKeycloakSessionFactory, ProviderLoader, 
>> ProviderManager, DefaultProviderLoader). 

Ok, let's include it

>> It would be better if it added additional liquibase changelog, rather than rely on modifying 
>> jpa-changelog-master.xml. Should be fairly easy to do.

ATM the PR adds possibility to use custom changelog file instead of jpa-changelog-master.xml. My question was, if it's possible to include the original file jpa-changelog-master.xml from the newly configured "master" file? In details there are 2 possible cases:

1) If you want to include custom changelog file with just 2 new tables and you can include jpa-changelog-master.xml from your new "master" file, you won't need to update your changelog every time when you update to newer Keycloak. That will be good IMO.

2) On the other hand, if your new "master" changelog file can't include jpa-changelog-master and instead it needs to include all the "partial" keycloak changelog files (jpa-changelog-1.0.0.xml , jpa-changelog-1.1.0.xml etc) together with your new changelog file for 2 tables, you will need to update your new master changelog file every time when you update to new keycloak version. That's not very ideal IMO as we will need to update example too during each release. 

@ewjmulder did you had a chance to check if case 1 works or did you just manually include all the "partial" changelogs in your new master changelog?

stianst commented:
It shouldn't change the master change log file. That should always be
included and hard coded in our code as it is currently in master. That
should still be what's used for built in entities.

The option should be to also include a separate master change log files.
That should be used for additional custom entities.

The built-in and custom entities should be in separate tables. Otherwise
there will be a migration nightmare.
On 22 Mar 2016 10:02, "Marek Posolda" <notifications@github.com> wrote:

> I'm happy with the changes inside classes DefaultKeycloakSessionFactory,
> ProviderLoader,
> ProviderManager, DefaultProviderLoader).
>
> Ok, let's include it
>
> It would be better if it added additional liquibase changelog, rather than
> rely on modifying
> jpa-changelog-master.xml. Should be fairly easy to do.
>
> ATM the PR adds possibility to use custom changelog file instead of
> jpa-changelog-master.xml. My question was, if it's possible to include the
> original file jpa-changelog-master.xml from the newly configured "master"
> file? In details there are 2 possible cases:
>
> 1) If you want to include custom changelog file with just 2 new tables and
> you can include jpa-changelog-master.xml from your new "master" file, you
> won't need to update your changelog every time when you update to newer
> Keycloak. That will be good IMO.
>
> 2) On the other hand, if your new "master" changelog file can't include
> jpa-changelog-master and instead it needs to include all the "partial"
> keycloak changelog files (jpa-changelog-1.0.0.xml , jpa-changelog-1.1.0.xml
> etc) together with your new changelog file for 2 tables, you will need to
> update your new master changelog file every time when you update to new
> keycloak version. That's not very ideal IMO as we will need to update
> example too during each release.
>
> @ewjmulder <https://github.com/ewjmulder> did you had a chance to check
> if case 1 works or did you just manually include all the "partial"
> changelogs in your new master changelog?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/keycloak/keycloak/pull/2219#issuecomment-199728677>
>

ewjmulder commented:
Thanks for the review. To answer your questions:
* Regarding the SPI loading: We did not include this for the REST extension. We use it to load our own service classes from the KeycloakSession object, eg `session.getProvider(OrganizationService.class);` instead of creating a new service instance 'manually' every time by calling it's constructor and trying to pass the reference around. Then we can use the session in the service to acquire for instance the entity manager (to CRUD our custom JPA entities) or the realm from the session context. I think this is an essential part of the PR, because it enables you to really integrate with the Keycloak application.
* As for the Liquibase changelog: I totally agree the Keycloak master change log chould not be touched and always included 'as is'. Indeed we used the solution nr 1 you described: We have our own master changelog file and it includes the Keycloak master changelog file and our own additional changelog file(s). Also we do not touch any of the build in entities or tables, to prevent migration trouble. You could add this as 'the way to do it' in the documentation. Even better might be to organize the code such that Liquibase will always execute the Keycloak master changelog and optionally executes another custom (master) changelog, if configured. But since the Liquibase constructor only accepts one String that points to a change log file (http://www.liquibase.org/javadoc/liquibase/Liquibase.html) I do not see an easy solution for this. We could consider constructing a ResourceAccessor that resolves a dynamically generated changelog file that includes the Keycloak master changelog and another custom one, if configured. But we could also settle with a specific instruction in the documentation and a warning that things might break if you do otherwise. What do you think?
* One important thing to mention here is that the Keycloak code contains a check that tries to 'outsmart' Liquibase in my opinion: getCurrentVersionSql. This is used to not trigger Liquibase if the latest version is found to be up to date. This logic now no longer applies (when using a custom changeset), since the custom changeset version number will never match the latest Keycloak version number. It doesn't hurt either, because then Liquibase will be triggered and it will see for itself the DB is up to date. But you might consider removing that check. Also because it uses plain SQL that is not universally valid, because it uses the 'LIMIT' keyword which is not supported by Sybase for example (the DBMS we use).
* Missing testing: I will create an example and use that both as an integration test and a piece of documentation. Should it be included in this PR or a separate PR?
* Missing documentation: what do you mean? Explaining the new functionality in the Keycloak docs I guess? I did not put effort in this yet, because I first liked to know if the PR would be accepted. I can create this of course. Again: same PR or different one?

stianst commented:
Could an option be to run Liquibase first for Keycloak internals, then run
it again for custom change log?

There was some reason I added getCurrentVersionSql. Think it was related to
us not using Liquibase initially. Marek could you look into that?
On 23 Mar 2016 10:06, "Erik Mulder" <notifications@github.com> wrote:

> Thanks for the review. To answer your questions:
>
>    - Regarding the SPI loading: We did not include this for the REST
>    extension. We use it to load our own service classes from the
>    KeycloakSession object, eg
>    session.getProvider(OrganizationService.class); instead of creating a
>    new service instance 'manually' every time by calling it's constructor and
>    trying to pass the reference around. Then we can use the session in the
>    service to acquire for instance the entity manager (to CRUD our custom JPA
>    entities) or the realm from the session context. I think this is an
>    essential part of the PR, because it enables you to really integrate with
>    the Keycloak application.
>    - As for the Liquibase changelog: I totally agree the Keycloak master
>    change log chould not be touched and always included 'as is'. Indeed we
>    used the solution nr 1 you described: We have our own master changelog file
>    and it includes the Keycloak master changelog file and our own additional
>    changelog file(s). Also we do not touch any of the build in entities or
>    tables, to prevent migration trouble. You could add this as 'the way to do
>    it' in the documentation. Even better might be to organize the code such
>    that Liquibase will always execute the Keycloak master changelog and
>    optionally executes another custom (master) changelog, if configured. But
>    since the Liquibase constructor only accepts one String that points to a
>    change log file (
>    http://www.liquibase.org/javadoc/liquibase/Liquibase.html) I do not
>    see an easy solution for this. We could consider constructing a
>    ResourceAccessor that resolves a dynam! ically ge nerated changelog file
>    that includes the Keycloak master changelog and another custom one, if
>    configured. But we could also settle with a specific instruction in the
>    documentation and a warning that things might break if you do otherwise.
>    What do you think?
>    - One important thing to mention here is that the Keycloak code
>    contains a check that tries to 'outsmart' Liquibase in my opinion:
>    getCurrentVersionSql. This is used to not trigger Liquibase if the latest
>    version is found to be up to date. This logic now no longer applies (when
>    using a custom changeset), since the custom changeset version number will
>    never match the latest Keycloak version number. It doesn't hurt either,
>    because then Liquibase will be triggered and it will see for itself the DB
>    is up to date. But you might consider removing that check. Also because it
>    uses plain SQL that is not universally valid, because it uses the 'LIMIT'
>    keyword which is not supported by Sybase for example (the DBMS we use).
>    - Missing testing: I will create an example and use that both as an
>    integration test and a piece of documentation. Should it be included in
>    this PR or a separate PR?
>    - Missing documentation: what do you mean? Explaining the new
>    functionality in the Keycloak docs I guess? I did not put effort in this
>    yet, because I first liked to know if the PR would be accepted. I can
>    create this of course. Again: same PR or different one?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/keycloak/keycloak/pull/2219#issuecomment-200277939>
>

mposolda commented:
Yeah, seems we didn't have liquibase in very first Keycloak "1.0.0.Final" version. I think that's the  reason for workarounds in DefaultJpaConnectionProviderFactory (calling getCurrentVersionSql) and in LiquibaseJpaUpdaterProvider (calling SELECT id from REALM and checking if something is in the REALM table). My opinion is to drop support migrating from prehistoric first version "1.0.0.Final" to 2.0 and remove workarounds. If someone really needs to upgrade from this version, he has the option to migrate first from "1.0.0.Final" to "1.9.1" (or anything newer) and then later from "1.9.1" to newer version.

Eric, can you create JIRA for this with fix version "2.0.CR1" and link it with your JIRA KEYCLOAK-2474?

- For docs, testing and example you can do another commit in this PR and add it all here. For documentation, I suggest to do another chapter like "Extending Keycloak" or something like that and maybe add the links to this chapter from the chapter 3 (Database configuration) and from the chapter 4 (Providers and SPI), so someone who is reading basic stuff in these chapter is aware that for more "advanced" stuff he can go to your chapter.

ewjmulder commented:
Ok, sounds good. I will create the JIRA ticket for removing the old pre-Liquibase workarounds. And I will update this PR with tests, documentation and an example. Furthermore, I'll look into a way to make the Liquibase extension more robust, so extensions can't skip the Keycloak master changelog. I'm not sure how soon I can do all of this, but I read on the mailing list the 2.0 is not due for some time right?

ewjmulder commented:
Ok, issue KEYCLOAK-2714 created for removing the old pre-Liquibase workaround, set to fix version 2.0.0.CR1 and linked to KEYCLOAK-2474.

mposolda commented:
Thanks! Yes, 2.0 don't have any exact date, first release could be around summer/autumn

stianst commented:
FIY I stole some of the code that creates the EntityManagerFactory. This was so persistence.xml can be loaded from a module marked as private.

ewjmulder commented:
Ok sure, what impact does that have on this PR?

stianst commented:
Not much, I moved the createEntityManager into JMSUtils as it can be useful for others that want to create their own EntityManagerFactory and have peristence.xml in a module. It was also changed slightly as I didn't need the external entities now.

stianst commented:
Looks like a cleaner way to create EMF would be:
    p.put("hibernate.classLoaders", classLoaders);
    …
    emf = Persistence.createEntityManagerFactory("acmeEntities", p);

ewjmulder commented:
Yes, but only when you just want to scan other modules for persistence.xml and Entity classes, which I believe in your case is enough. If you want to modify the persistence unit, in case of this PR, you do need the more elaborate code I wrote so you can add entities.

stianst commented:
Of course. For a second I was thinking all you wanted was to add class
loader, but that's obviously not enough.
On 7 Apr 2016 09:55, "Erik Mulder" <notifications@github.com> wrote:

> Yes, but only when you just want to scan other modules for persistence.xml
> and Entity classes, which I believe in your case is enough. If you want to
> modify the persistence unit, in case of this PR, you do need the more
> elaborate code I wrote so you can add entities.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/keycloak/keycloak/pull/2219#issuecomment-206746725>
>

stianst commented:
Assigned @mposolda to review and refactor this before merging.

ewjmulder commented:
Sorry I left this open for so long. My company is happy now with the current extensions and the issue to complete this PR is waiting in the backlog for ages...
Maybe I can squeeze it in this week if you guys say the 2.0.0.CR1 is really coming up now :)

mposolda commented:
fyi. I've fixed the JIRA KEYCLOAK-2714 for removing pre-liquibase workarounds. It's in master now.

mposolda commented:
Yes, the 2.0.0.CR1 is coming up soon now :) So if you have opportunity to finish, it will be cool. The things, which are potential remaining todo are:

1) Rebase this with latest master.

2) Change PR to use existing JpaUtils.createEntityManagerFactory instead of adding new private method DefaultJpaConnectionProviderFactory.createEntityManagerFactory . That's because Stian forked your code in the meantime after you sent this PR, you know ;) This probably means that JpaUtils.createEntityManagerFactory will need to be slightly changed to support adding custom JPA entities?

3) More flexible way of "inject" custom liquibase changelogs for migration new entities. The idea is, that you will be able to add your own changelog even without the need to create custom subclass and override DefaultLiquibaseConnectionProvider.getChangelogLocation . So maybe if you can extend your JpaEntityProvider and add some additional method (or additional methods) like:

boolean requireMigration();

String getChangelogLocation();

Then LiquibaseJpaUpdaterProvider.update will first invoke the migration of keycloak master changelog (like it's now) and then it will retrieve changelogs from all JpaEntityProvider, which support migration and call update for them too? So update of keycloak master changelog is always done first and then remaining changelogs are done after. Will it work for you?

Maybe you will need to do some changes in both LiquibaseJpaUpdaterProvider and LiquibaseDefaultConnectionProvider to supports this (especially the ability, so liquibase is able to find changelogs from the custom JAR with your provider).


4) Documentation, example, tests - you can add them in separate PR. Is it working for you?

WDYT? Will you have time to followup this or next week?

ewjmulder commented:
Thanks for making an overview of the remaining tasks. I agree with your summary. I really want to get this thing done now, so I will make an effort to get this done next week at the latest. I cannot promise it'll all be done though, since it's a very busy time for me personally as well. I will work on the tasks in the order you wrote them down, so maybe if I'm able to finish points 1, 2 and 3 I can deliver point 4 a few days later?

mposolda commented:
That will be cool, thanks! Sure, we can sync next week or just write when you have 1,2,3. Anyway 4 can be done in separate PR.

ewjmulder committed:
KEYCLOAK-2474: Resolve conflicts

Merge branch 'master' of https://github.com/keycloak/keycloak

Conflicts:
	model/jpa/src/main/java/org/keycloak/connections/jpa/DefaultJpaConnectionProviderFactory.java
	services/src/main/java/org/keycloak/services/DefaultKeycloakSessionFactory.java

ewjmulder committed:
KEYCLOAK-2474: Bugfix: clash between local and class property name.

mposolda commented:
Hey Eric, so am I understand correctly that you already did 1+2 , but 3+4 is still TODO? Would you have time for more stuff for today or tomorrow? Or should I rather finish remaining stuff?

ewjmulder commented:
Hi Marek, indeed I finished 1 + 2. For 3 I do have a possible solution ready, but I did not have time to test it yet. Point 4 I didn't do yet, but I also don't have time to finish this week. Is it also ok to have 3 + 4 done on monday morning? I cannot spend time on it at my work anymore, but I will have some time in the weekend. Or what do you mean with finish remaining stuff?

mposolda commented:
yes, I meant just points 3+4 when talking about "remaining stuff". So we can sync on Monday morning and I can continue then? 

Btv. if you have some other better plan for weekend then coding, then just let me now :) It's not a problem for me to finish the remaining parts.  Thanks again for your work on this PR!

ewjmulder commented:
Ok, sure, if you have time left, you can work on it as well. Maybe I can review it then? How about I commit what I came up with for point 3 and you pick it up from there? Next weekend is my first holiday weekend, so a bit less coding wouldn't hurt ;)

ewjmulder committed:
KEYCLOAK-2474: Work in progress: load extra changesets dynamically.

ewjmulder committed:
Undo commit of .metadata

ewjmulder committed:
Merge branch 'master' of https://github.com/keycloak/keycloak

ewjmulder commented:
I just committed my untested idea for a solution for point 3 (sorry for the bogus .metadata commit, you can remove those from the commit history if you want).

What we whould want to tell Liquibase is to load several different changelog files from different locations. Unfortunately, I do not see an option for this in the Liquibase API. Also, I don't think you can trigger Liquibase multiple times on the same database with different changelog files, because it will be confused about the missing ones that are in the changelog table but not in the file. I did not test this idea though, so it might be interesting to try that out. That would make it somewhat simpler,

So the solution I came up with is to generate a changelog 'file' dynamically, composed of all the changeset of the different changelog files. I did check the Liquibase source code and this is essentially what they do themselves if they encounter an 'include' in the XML. Also, as you suggested, there is an extra method on the JpaEntityProvider: getChangelogLocation(). I did not make it 'optional' with an extra boolean method, because it does not make sense to have extra entities but no added changelog. Also I refactored the proxy classloader to be a more general solution to both loading the entities and the changelog file. As said before, I did not test this yet. One last detail about this is that you must be able to load the changelog file from the same classpath (so same added module) as the entities. But this seems like a very reasonable requirement.

Let me know what you think and if you can continue with this and if I can help you with anything.

ewjmulder committed:
Undo commit of .recommenders

mposolda commented:
I agree that first idea you mentioned will be somewhat simpler and to me personally looks better and more clean. Liquibase has method "setDatabaseChangeLogTableName" on class liquibase.database.Database . So seems it's not an issue to use different tables for different liquibase changelogs? The name of changelog table could be possibly deduced from the ID of JpaEntityProvider (ie. if FooJpaEntityProvider has id "foo-bar" then table name will be DATABASECHANGELOG_FOO_BAR ?

So then after calling liquibase.update for main keycloak changelog master, it can just iterate through all JpaEntityProvider and call liquibase.update for each?

mposolda commented:
btv. if you don't have time, I can try to check the idea above and test everything and also continue with remaining stuff (example, test, docs)

ewjmulder commented:
Ah, didn't know that you could use different changelog tables. I'm not really familiar with those kind of details of Liquibase. Then yes indeed, that is a much better solution. Both in terms of a cleaner coded solution and in a nice separation of changelog tables in the database. I like your idea of using the provider id as postfix of the changelog table name, prevents extra annoying config.

Yes please, if you could persue this idea further and make a working solution out of it, that'll be great. As for the other remaining stuff: what kind of example did you have in mind? I can at least describe the extension we made, how and why. Maybe not with our original code, but to get an idea of the possibilities. So if you could do the testing and documentation, I'll write some text about our use case and if I have time a little code and config of a small toy problem to demonstrate how it works.

mposolda commented:
TBH I am not sure about best approach for example. We are usually trying to have examples as easy as possible and focus on single thing. But on the other hand, if you want custom JPA entities or custom SPI, then you usually need custom REST endpoint as well, right? And we already have example for custom REST endpoint ( examples/providers/rest ).

So I can think about those approaches:
a) Keep current "rest" example and add new bigger example for custom REST endpoint + custom SPI + custom entity
b) Extend current example and add custom SPI + custom entity to it
c) Keep current "rest" example and add 2 new smaller examples like: 
   c.1) custom REST + custom SPI 
   c.2) custom REST + custom entity
d) Remove current "rest" example and add 2 new examples like in case c.

Not sure what is best approach. Maybe (a) ? Do you have any other idea? 
@stianst what do you think?

ewjmulder commented:
I'd say a) as well. Custom REST is a feature on it's own, so it deserves a separate example. But domain extension is only useful if you combine it with at least an extra changelog and custom REST endpoints and probably custom SPI.
Custom SPI has no use on it's own, but only combined with custom REST endpoints. So if you want to be fully complete you whould have:
1. Just REST example (already done)
2. Rest + custom SPI
3. Custom JPA + Liquibase + REST with mentioning SPI as additional useful extra.

But maybe you can skip 2. or add it as sidenote in 1.

The tricky thing here is where to explain a feature that has no usefulness just on it's own :)

mposolda commented:
ok, so if you have time, you can try to add new example for Custom JPA + Liquibase + REST + SPI into "examples/providers" ? Feel free to add it to this PR. I will continue on Monday with remaining parts (investigate separate changelogs + tests + docs + finish example)

ewjmulder commented:
Ok, perfect, I will start writing the combined example.

ewjmulder committed:
KEYCLOAK-2474: Added Example project for Domain Extension (WIP)

ewjmulder commented:
I've added an example project for the domain extension, including custom JPA entities, a custom changelog, a custom service SPI and a custom REST extension. It is based on our use case at Docdata, where we use Keycloak to authenticate our backoffice client users. Every keycloak user is assigned to a company and every company can have multiple regions. A user can have a (keycloak) role assigned at the level of a region. The example is not fully complete yet as far as CRUD actions at the REST and service level. I think it doesn't need to be fully functional as you would make it for a real life use case, just complete enough to see the whole picture and I'm almost there yet.

Just two issues I have now:
1. I cannot test it, since the Liquibase custom changelog loading solution is not in place yet.
2. The REST extension SPI is hooked into the RealmResource, while for the real life use case we do require the RealmAdminResource. Because we do want only an authenticated user to be able to for instance CRUD a company. Pedro initially made 2 hookup points: one for RealmResource and one for RealmAdminResource and that is also what I initially included in this PR. But I see Stian only included the RealmResource extension. I think for the domain extension functionality we really need the RealmAdminResource REST hook as well. I can make a PR for that as well or you can do it yourself in line with the RealmResource hook already in place, if you agree to include it. Otherwise I can still make an example using just the RealmResource, but it would be a lot less usable. If you'd like me to elaborate on this point, please let me know.

mposolda commented:
Closed as I've added another PR https://github.com/keycloak/keycloak/pull/2951 where I cherry-picked your commits and added other changes. My only changes were:
- Custom changelog table per each changelog and running each "partial" update separately as we discussed above.
- Simplified example a bit (for example I kept just the Company entity for the sake of simplicity). I've also added one secured endpoint to example, which requires authenticated user in role "admin" . Only thing you need to have your REST endpoint secured is to use "new AppAuthManager().authenticateBearerToken" and then you can check roles programatically as you want. Hence the RealmAdminResource extension is also not needed.
- Added automated test

The remaining work is still documentation. Eric, if you have some additional feedback let me know. It looks we will have 2.0.0.CR1 release quite soon :)

ewjmulder commented:
Thanks for the reference to the AppAuthManager.authenticateBearerToken, that does indeed provide the information we need. So I agree the RealmAdminResource is no longer needed.
